### ISSUE TITLE
Adding support for the Xbox Elite Controller 2

### ISSUE THREAD

**Issue by RicardoR12534** [ndeadly/MissionControl#3]
Since the Elite Controller 2 has native Bluetooth support and the controller is *pretty* similar to the regular Xbox Controller (don't quote me on that) you could add support for it?

**Comment by ndeadly** [comment]
I didn't realise there was a Bluetooth version of the Elite Controller. It's probably just a matter of whitelisting the hardware ID.

If you have one could you attempt to pair it via the Controllers->Change Grip/Order screen, then use btdb.nro provided with the release to dump the Bluetooth database and post a screenshot here.

**Comment by shyiro** [comment]
I've got the Elite Controller 2, and we can't pair him, so we can't get any enter on the Bluetooth database

**Comment by ndeadly** [comment]
@shyiro even though the controller might not appear to connect to the switch, it should be possible to complete the pairing process with the console and get it stored in the database. It will only show up as a connected controller if a handler gets installed for it, which requires a match against the hardware ID to determine the controller type.

**Comment by shyiro** [comment]
ok i see, but i've try a lot of time to pairing him, but the controller keep blink and no enter in the database for him. i don't know if it's linked, but, when i try to pairing him, and go to btdb.nro for see if he is here, sometime i get a crash of atmosphere.

**Comment by ndeadly** [comment]
[MissionControl-0.1.0-xbox-one-elite-2.zip](https://github.com/ndeadly/MissionControl/files/5138823/MissionControl-0.1.0-xbox-one-elite-2.zip)
I've whitelisted a hardware id I found online that appears to be the Elite Controller v2. See if this build makes any difference for you. Also might be worth making sure your controller firmware is up to date

**Comment by shyiro** [comment]
doesn't work, Atmosphere keep crashing a few seconds after i try to pair him

**Comment by ndeadly** [comment]
Hmm, that's odd. Can you post the fatal report here so that I can see what's crashing and where?

**Comment by shyiro** [comment]
![photo_2020-08-28_09-23-34](https://user-images.githubusercontent.com/22276228/91539606-e0037a80-e919-11ea-939f-161c467ce78d.jpg)

the crash come randomly after i try to pair him, if i stay in the controller menu, or in the hbmenu, anywhere i go, it's crashed a few seconds after try to pair.

**Comment by ndeadly** [comment]
Thanks. I'll check it out when I have some time.

For future reference, it's much better if you post the report log file that gets written to your SD card under `atmosphere\fatal_reports`. That way I can use automated tools to inspect it rather than having to key in the addresses manually ;)

**Comment by shyiro** [comment]
oh sorry, i didn't know that ^^

[01598542087_010000000000002a.log](https://github.com/ndeadly/MissionControl/files/5141140/01598542087_010000000000002a.log)

**Comment by victorgneves** [comment]
> [MissionControl-0.1.0-xbox-one-elite-2.zip](https://github.com/ndeadly/MissionControl/files/5138823/MissionControl-0.1.0-xbox-one-elite-2.zip)
> I've whitelisted a hardware id I found online that appears to be the Elite Controller v2. See if this build makes any difference for you. Also might be worth making sure your controller firmware is up to date

This modification was very useful.
Friend of mine only managed to connect his control with this version, and it is not an elite control.
It is a normal Xbox One controller.
Maybe it's a slightly different version.

**Comment by ndeadly** [comment]
@victorgneves that's strange, is it an official Microsoft controller? As far as I can remember, the only change I made for this build was adding the ID `{0x045e, 0x0b00}` to the whitelist for Xbox One controllers. Can your friend dump their Bluetooth database with `btdb.nro` and post it here?

**Comment by ndeadly** [comment]
@shyiro try this build with logging enabled, and post the resulting `btdrv-mitm.log` from your SD root here. Note that the log gets wiped on console boot, so if your switch crashes you will need to either pull the SD out and read it in a computer, or mount it from hekate to read it before booting.

[MissionControl-0.1.0-debug-logging.zip](https://github.com/ndeadly/MissionControl/files/5145442/MissionControl-0.1.0-debug-logging.zip)

**Comment by PabloPerezRodriguez** [comment]
I've got the same controller, and get (seemingly) the same crash (I go to change grip/order, try to pair my controller, and when I try to return to the main menu, it crashes).
Here's my atmosphere fatal log 
[01549653030_010000000000002a.log](https://github.com/ndeadly/MissionControl/files/5162069/01549653030_010000000000002a.log)
And here's my btdrv-mitm.log (from the last build you sent)
[btdrv-mitm.log](https://github.com/ndeadly/MissionControl/files/5162071/btdrv-mitm.log)

**Comment by ndeadly** [comment]
Thanks @PabloPerezRodriguez . Unfortunately, nothing too interesting in those logs. It looks like it might just try to pair twice and then crash. Could you try with this new build with more verbose logging of the pairing process?

[MissionControl-0.1.0-debug-logging2.zip](https://github.com/ndeadly/MissionControl/files/5163809/MissionControl-0.1.0-debug-logging2.zip)

**Comment by elad0elad** [comment]
First thank you for the great work, second I join the request to add the elite v2 to the supported list please

**Comment by ndeadly** [comment]
@elad0elad you could try running the build I've posted above and post your log. I can't add support if I can't figure out why it won't connect. I don't own this controller myself.

**Comment by PabloPerezRodriguez** [comment]
Sorry for the delay, my SD Card got corrupted and I was trying to recover my data (didnt work).

I tried again with the new build. These are the logs.
[btdrv-mitm.log](https://github.com/ndeadly/MissionControl/files/5179686/btdrv-mitm.log)
[01550003809_010000000000002a.log](https://github.com/ndeadly/MissionControl/files/5179687/01550003809_010000000000002a.log)
I also noticed Atmosphere creates crashdumps, not sure if you want that, I'm not uploading it here because I don't know if it contains console specific info.

**Comment by elad0elad** [comment]
> @elad0elad you could try running the build I've posted above and post your log. I can't add support if I can't figure out why it won't connect. I don't own this controller myself.

I've been trying this and still don't work friend, there is any thing I can do to help the progress?

**Comment by ndeadly** [comment]
Sorry guys, I'm busy preparing some other things for the next release at the moment. I want to knock you up a homebrew application to do the bluetooth pairing myself and see if we can at least avoid the crash for further debugging. Give me a day or two.

@PabloPerezRodriguez thanks for the logs. Looks like you're getting authentication failure status events before the crash. Not exactly sure what could cause this. Reading around a bit it kind of seems a bit like the QC on the elite controller is worse than the standard ones that come with the xbox. I don't think you've mentioned whether your controller is on the latest firmware or not? Make sure it's up to date as it sounds like there have been improvements to the bluetooth function over time.

You can upload the Atmosphere crashdumps too if you want. At the moment I don't really need them as the stack trace is enough, but they might prove useful if I end up having to dive deeper into the disassembly to figure out what's going wrong.

**Comment by PabloPerezRodriguez** [comment]
It is on the latest firmware. And don't worry about time, I can wait!

**Comment by elad0elad** [comment]
There is an option to add the controller by enter the Mac address manually?

**Comment by ndeadly** [comment]
@elad0elad no, the pairing process has to complete with the controller to exchange link keys with the console.

**Comment by tomlube** [comment]
I can't even get my Elite 2 to connect at all :(

**Comment by symalekos** [comment]
i tried so may times to pair my elite 2 controller with the console but no luck. The controller is blinking like its searching to pair but no luck.Also i try the btdb.nro program but it says no paired bluetooth devises

**Comment by ndeadly** [comment]
Hey guys, sorry it's taken so long. Other things have got in the way.

Here's a little homebrew app I made to reimplement the pairing process. It's not perfect and may have some bugs, but hopefully it won't crash if you try to pair the elite controller. See if you can manage to get it to do anything interesting. A successful pairing will show the device info in green. It creates a log at `sdmc://bluetooth_pairing.log` that you can upload here for me to take a look at.

[bluetooth_pair.zip](https://github.com/ndeadly/MissionControl/files/5252776/bluetooth_pair.zip)

**Comment by Dallas-J** [comment]
[bluetooth_pairing.log](https://github.com/ndeadly/MissionControl/files/5253305/bluetooth_pairing.log)

Here's a log - I left it running for a bit just in case some logs were different than others. Thanks for doing all this man!

**Comment by ndeadly** [comment]
@Dallas-J thanks for the log. I assume this didn't crash on you then by the looks of that log? I probably should have timestamped it.

Looks as though it's continually failing the authentication step. Here are a few versions I built tweaking the parameters to the authentication response. I don't see why this controller should be different to any other, but I guess it can't hurt to try. Let me know if you find success with any of these.

[bluetooth_pair_A.zip](https://github.com/ndeadly/MissionControl/files/5254431/bluetooth_pair_A.zip)
[bluetooth_pair_B.zip](https://github.com/ndeadly/MissionControl/files/5254432/bluetooth_pair_B.zip)
[bluetooth_pair_C.zip](https://github.com/ndeadly/MissionControl/files/5254434/bluetooth_pair_C.zip)
[bluetooth_pair_D.zip](https://github.com/ndeadly/MissionControl/files/5254435/bluetooth_pair_D.zip)

**Comment by elad0elad** [comment]
[bluetooth_pairing.log](https://github.com/ndeadly/MissionControl/files/5258311/bluetooth_pairing.log)
here's my log,
allready try all 4 version and the one thet you upload earlier, nothing change

**Comment by ndeadly** [comment]
@elad0elad damn, this controller really doesn't want to work.

Are any of you guys able to pair this controller on linux and share the pairing info? This might help understand if there's anything different to a regular Xbox One controller. I'm also thinking a workaround might be to pair the controller under switch linux and inject the info into the database under HOS.

**Comment by Dallas-J** [comment]
[bluetoothctl.log](https://github.com/ndeadly/MissionControl/files/5273023/bluetoothctl.log)

I'm uploading this from switchroot - is this what you're looking for? I had just disconnected the controller so it failed to pair at first.

**Comment by kylefmohr** [comment]
Is this what you're looking for @ndeadly ? Let me know if I can provide additional information

[hcitool info.log](https://github.com/ndeadly/MissionControl/files/5273392/hcitool.info.log)

**Comment by ndeadly** [comment]
@Dallas-J , @kylefmohr sorry if there was confusion from me asking for logs earlier. That was mostly for diagnosing what was going on under HOS. What I'm actually after from linux is the equivalent to what would get stored in the HOS bluetooth database. I don't know if it's the same across most distros, but under ubuntu this can be found at `/var/lib/bluetooth/<host_address>/<client_address>/info`. My hope is that I might be able to find all the necessary info in here to translate to the switch database format and inject it into the database under HOS. I'd still like to figure out why the pairing fails on HOS, but this could act as a workaround for you guys in the meantime. The ability to do this would also be useful for people running switch linux or emuMMC who want to synchronise their pairing databases.

**Comment by kylefmohr** [comment]
Like this?

root@kyle:/var/lib/bluetooth/98:B6:E9:AA:35:57/98:7A:14:5B:35:BF# cat info
[General]
Name=Xbox Elite Wireless Controller
Class=0x000508
SupportedTechnologies=BR/EDR;
Trusted=true
Blocked=false

[DeviceID]
Source=2
Vendor=1118
Product=2821
Version=1032

[LinkKey]
Key=16E555BC220168E6E8C85E9C6D55A716
Type=4
PINLength=0

**Comment by ndeadly** [comment]
@kylefmohr yeah that's perfect. I'll try to finish writing some tools to try this out after I'm done with work today.

**Comment by kylefmohr** [comment]
Hi @ndeadly, just wanted to see if there was any additional information I could provide to assist with adding this controller to MissionControl. 

Thanks for all of your hard work so far. I’d be happy to send another donation once this is added 🙂

**Comment by ndeadly** [comment]
Hey, sorry for the lack of updates. I've written a python script to dump the database from L4T and have some basic functionality to import it back into HOS, but it doesn't seem to work in all cases for some reason. In particular I couldn't get my regular Xbox One controller to be recognised. I need to do more research. I will try to dedicate a bit more time to look into it this week. It's hard when I don't own the controller

You could check for me if any of the changes I made in 0.3.0 have made any difference to the pairing situation if you like. Thanks for your support :)

**Comment by kylefmohr** [comment]
No worries! I will try 0.3.0 tonight and let you know if there's any noticeable change. I'd be happy to test the L4T and HOS script for you once it's ready, if that would help.

**Comment by kylefmohr** [comment]
No change on 0.3.0 unfortunately, still not detected.

**Comment by ndeadly** [comment]
@kylefmohr yeah not all that surprising. Just wishful thinking on my part that the thread priority bug I solved in 0.3.0 may have been to blame.

**Comment by Robert2282** [comment]
Hey I was wondering if you made any progress on the elite controller.  I'm guessing the controller does some strange stuff when paring because I also have the mayflash switch Bluetooth adapter and I think its got a unique pairing process as it took around 15-30 seconds to pair.  Also here a quote from their pairing guide for the elite controller

- It may take a longer time to connect your Xbox Elite 2 and Magic-S adapter. And the connection may be disconnected shortly (The lights on the controller and the adapter flash) and then connect again during the Bluetooth pairing process. Please wait patiently until  the LED indicators on the controller and the adapter stop flashing.

I hope this helps

**Comment by kintsugi** [comment]
> Hey I was wondering if you made any progress on the elite controller. I'm guessing the controller does some strange stuff when paring because I also have the mayflash switch Bluetooth adapter and I think its got a unique pairing process as it took around 15-30 seconds to pair. Also here a quote from their pairing guide for the elite controller
> 
>     * It may take a longer time to connect your Xbox Elite 2 and Magic-S adapter. And the connection may be disconnected shortly (The lights on the controller and the adapter flash) and then connect again during the Bluetooth pairing process. Please wait patiently until  the LED indicators on the controller and the adapter stop flashing.
> 
> 
> I hope this helps

To add to this line of thinking, I use [switch up adapter](https://www.switch-up.ca/) for it's bluetooth capabilities and it pairs without delay with xbox elite controller 2 to the adapter, and then a couple second delay before the controller led stops flashing.

**Comment by goodwind** [comment]
Good day! Is there any progress in pairing Elite v2 Controller? If you need any assistance I could help by launching some tools or scripts and providing you it's output.

**Comment by ndeadly** [comment]
Hey, sorry, no major updates. I've been sick the past week or so, so not a lot of dev work being done. I'm slowly working towards having a gui application that will allow exporting and importing the controller database amongst other things. It _**might**_ be possible to use this to restore the pairing info from l4t, but it's also possible that controller is being rejected by the console during connection for some other reason. It might require patching of the bluetooth service. This is a difficult problem to solve without access to the controller to be able to live debug the connection process.

**Comment by kylefmohr** [comment]
Is this something we could crowdfund for you? It's a very nice controller, albeit pricey. I'm in for $10, anybody else? And is this something you're interested in @ndeadly ?

Failing that, I have a fair amount of Bluetooth receivers, access to any VM you might need, and at least adequate knowledge of how all these things work (though I know almost nothing about the nitty gritty of Bluetooth connections). Would you want to remote into a VM I host and you could instruct me to put the controller in pairing/move files to the Switch as needed? Just a thought, I'm willing to help however I can.

**Comment by ndeadly** [comment]
@kylefmohr if people wanted to crowdfund the controller for me so that I can dig deeper I'd happily accept. It needs to be made clear though that this will not necessarily guarantee a solution. The last thing I want is people feeling that I owe them any results because of it.

While I appreciate the offer of remote access, I don't think it's really practical. A lot of this work can require staring at decompiler output for hours and messing around in gdb, but also rapid testing of new ideas and iterating upon them. The delay/overhead from having to communicate every little thing, plus needing you there the whole time on standby would quickly frustrate us both.

**Comment by kylefmohr** [comment]
> @kylefmohr if people wanted to crowdfund the controller for me so that I can dig deeper I'd happily accept. It needs to be made clear though that this will not necessarily guarantee a solution

Yes, absolutely. Something I was taking for granted, but it's definitely a good idea to mention. Chipping in != guaranteed support for the controller (or any guarantees, really)

> A lot of this work can require staring at decompiler output for hours and messing around in gdb, but also rapid testing of new ideas and iterating upon them.

That makes sense.

**Comment by goodwind** [comment]
> plus needing you there the whole time on standby would quickly frustrate us both.

I am totally agree, but if you will need to test something or make some dump don't hesitate to contact me. I can't afford to crowdfund something with money, but I could sacrifice some of my free time

**Comment by ndeadly** [comment]
Just thought of something else to try...

If someone with an Android phone could try and connect the controller to their phone with the HCI snoop log enabled that might provide us with some additional info. The Android Bluetooth stack is similar to the one used by the Switch.

**Comment by shyiro** [comment]
here is mine
[btsnoop_hci.log](https://github.com/ndeadly/MissionControl/files/5465963/btsnoop_hci.log)

**Comment by kintsugi** [comment]
@ndeadly This is a bit unorthodox, but on the off chance we live in the same area I have the xbox elite 1, 2, and razer wolverine ultimate controller (as well as a bottle of 99% isopropyl alcohol for your safety) that I am comfortable lending for science.

**Comment by ndeadly** [comment]
@kintsugi haha I admire your enthusiasm. If you happen to live in Southern Germany we can talk ;)

**Comment by kintsugi** [comment]
> @kintsugi haha I admire your enthusiasm. If you happen to live in Southern Germany we can talk ;)

Haha thanks! I live in San Francisco, CA, so unfortunately it won't work out 😪

Here's an idea: I have a desktop computer I use for kvm/libvirt, I could spin up a linux/windows/whatever image for you, passthrough a bluetooth adapter and connect the controller. Then I could follow your instructions to send input/press buttons. What do you think?

**Comment by ndeadly** [comment]
Anybody want to update their controller to the latest firmware and give this build a try? It's been brought to my attention there may have been some firmware bugfixes recently, and I was also sending an unnecessary command to the controller. Don't want to get anyone's hopes up, but there's a small chance one of those things could have been the cause of problems for the Elite v2.

[MissionControl-0.4.1-no-xbox-init.zip](https://github.com/ndeadly/MissionControl/files/5784087/MissionControl-0.4.1-no-xbox-init.zip)

**Comment by kylefmohr** [comment]
I just know it's gonna work, because I bought an 8bitdo like 12 hours ago lol

I will test this afternoon and let you know shortly

**Comment by kakra** [comment]
I don't like to drop into a non-technical discussion but this one may be important. All non-programmers: Please ignore. :-)

> Anybody want to update their controller to the latest firmware and give this build a try?

https://github.com/ndeadly/MissionControl/blob/a916172c07bf31e3af693f860e394af838aa0096/bluetooth-mitm/source/controllers/xbox_one_controller.cpp#L39

@ndeadly The XBE2 reports are larger (and in the older buggy firmware much larger) than the older controllers reports. Does the code handle this gracefully? What happens if the packet is shorter than what's expected (as in `report->data` is shorter than `XboxOneReportData`)? Would this end in an out of bounds access and crash MC?

For xpadneo, we've made these captures: https://github.com/atar-axis/xpadneo/blob/master/docs/descriptors/xbe2_unknown.md

It could help you finish support for this controller. Also, it seems we need to enable Bluetooth privacy extensions for a reliable connection. IDK if this is something MC needs to bother with, or if it can change BT behavior of the console.

I think it may be worth considering the report size to tell the protocol variants apart although the first few bytes are always the same purpose. It starts to change with the buttons (XB1S controllers may either report 10, 11 or 16 buttons, the 16-button mode uses a sparse bitmap and the report is one byte longer, the 10-button mode reports the guide button in report 0x02). XBE2 adds some more bytes (older firmware has a duplicate set of bytes), the additional bytes have information about trigger mode, paddels and profile switcher. The new (not yet supported) XBXS controller reports 12 buttons, and shifts some button bits compared to the older models.

**Comment by kylefmohr** [comment]
You can always drop in if you're gonna provide useful information like that!

@ndeadly unfortunately no luck on the build you just posted, the controller simply doesn't show up. To recap, I:

- Ensured XBE2 had the latest firmware
- Overwrote previous MissionControl release with the no-xbox-init version
- Rebooted, ensured the sysmodule was running, rebooted again
- Also power cycled the XBE2

Still nothing in the change grip/order menu. My switch is running FW 11.0.1 + Atmosphere version 0.16.2

Just posting as much info as I can in case I'm forgetting something stupid.

**Comment by kakra** [comment]
@ndeadly Would it help to capture a `btmon` or `wireshark` dump in Linux for the controller connecting/pairing? Then maybe pressing each button and axis along with a video of showing all the controller functions?

**Comment by kakra** [comment]
BTW: The long connection handshake is gone in Linux since the kernel supports Bluetooth privacy extensions properly and this feature has been enabled in the bluez configuration file. For the Switch console, this may however not apply or work completely different (different Bluetooth stack).

**Comment by ndeadly** [comment]
@kakra the console stores hid reports to the following structure, so unless the incoming report size exceeds 700 bytes it shouldn't be an issue. I reinterpret cast this structure to the appropriate controller report definition. Worst case is the controls are mapped incorrectly.

```
typedef struct {
    u16 size; 
    u8 data[0x2BC]; 
} BtdrvHidReport;
```

We don't even appear to be making it to the point the controller starts sending reports though. I believe it's failing at some point during the connection/pairing process. I can't really debug this however as I can't log anything going on inside the driver. I would need the controller to either remote debug the driver from my PC or capture the packets via some other means.

> Also, it seems we need to enable Bluetooth privacy extensions for a reliable connection. IDK if this is something MC needs to bother with, or if it can change BT behavior of the console.
Not sure on that one either. I don't have a lot of influence over the console Bluetooth other than a few SSP pairing settings. Anything else would likely need to be patched in. Do the privacy extensions apply for all Xbox One controllers or just the Elite? I don't have any connection issues with the regular one on the Switch.

> @ndeadly Would it help to capture a `btmon` or `wireshark` dump in Linux for the controller connecting/pairing? Then maybe pressing each button and axis along with a video of showing all the controller functions?

Not sure, couldn't hurt to take a look. I think I already had someone grab me an hci snoop log from android earlier in this thread, but I don't recall seeing anything particularly interesting. My main interest is what happens during connection or authentication that might be different from the regular Xbox One controller.

**Comment by ndeadly** [comment]
> You can always drop in if you're gonna provide useful information like that!
> 
> @ndeadly unfortunately no luck on the build you just posted, the controller simply doesn't show up. To recap, I:
> 
> * Ensured XBE2 had the latest firmware
> * Overwrote previous MissionControl release with the no-xbox-init version
> * Rebooted, ensured the sysmodule was running, rebooted again
> * Also power cycled the XBE2
> 
> Still nothing in the change grip/order menu. My switch is running FW 11.0.1 + Atmosphere version 0.16.2
> 
> Just posting as much info as I can in case I'm forgetting something stupid.

@kylefmohr not particularly surprising, but thanks for trying. Does the console crash at all anymore?

**Comment by kylefmohr** [comment]
I personally have never experienced crashes due to MissionControl

**Comment by kakra** [comment]
> Not sure on that one either. I don't have a lot of influence over the console Bluetooth other than a few SSP pairing settings. Anything else would likely need to be patched in. Do the privacy extensions apply for all Xbox One controllers or just the Elite? I don't have any connection issues with the regular one on the Switch.

I'm not sure if it affected the older models as I was already able to pair them without problems in Linux. But it changed connection behavior of the XBE2 very visibly: It's more or less immediate now compare to a 20 second long "frozen phase" before. In a long gaming session, all models stay connected much more reliable, tho that may be a placebo effect or purely subjective: I did turn off rumble around the same time which makes all controller models much more stable in Bluetooth mode anyways.

Multiple users of xpadneo reported that turning on privacy extensions fixed most problems for them, with XBE2 and especially the new XBXS model.

Compared to Android (which seems to use a Bluetooth stack very similar to the Switch console): There the controllers seemed to connect without problems since the beginning. But I don't remember if I tried the XBE2 controller. I probably did try, and if I did, I would remember if it had connection issues. So let's assume it connected without problems on XBE2.

I'll try to create a Bluetooth dump of pairing/authentication in the weekend and send it to you. Any suggestions which Linux capturing tool I should use for best results?

**Comment by kakra** [comment]
BTW: All the PID/VIDs of the models are here: https://github.com/atar-axis/xpadneo/blob/master/hid-xpadneo/etc-modprobe.d/xpadneo.conf

The XBE2 controller is 0x0B05, XBXS is 0x0B13.

This is for Bluetooth mode. The USB modes of the controller actually use different IDs which you could see here:
https://github.com/atar-axis/xpadneo/blob/master/hid-xpadneo/src/hid-xpadneo.c#L864 (and following lines)

The Bluetooth IDs are also in the code here:
https://github.com/atar-axis/xpadneo/blob/master/hid-xpadneo/src/hid-xpadneo.c#L1160

Once connected/authenticated in Bluetooth, there should actually really be no difference in handling the controller in the code (except for some differences in the HID reports). So if XBE2 does not connect, it's probably somewhere in the Bluetooth handshake phase.

**Comment by ndeadly** [comment]
> Multiple users of xpadneo reported that turning on privacy extensions fixed most problems for them, with XBE2 and especially the new XBXS model.

Have you got any further info on these privacy extensions? I might try to take a look into it.

> Compared to Android (which seems to use a Bluetooth stack very similar to the Switch console): There the controllers seemed to connect without problems since the beginning. But I don't remember if I tried the XBE2 controller. I probably did try, and if I did, I would remember if it had connection issues. So let's assume it connected without problems on XBE2.

Yes, the switch Bluetooth stack is very similar, it's why I thought the Android hci log might have been useful. I don't think I've heard of anyone having issues on Android either.

> I'll try to create a Bluetooth dump of pairing/authentication in the weekend and send it to you. Any suggestions which Linux capturing tool I should use for best results?

Thanks, that would be helpful. I'm not super familiar with which tools are available for this purpose on linux, I thought you might have a better idea. I haven't needed to see the low level packets before so I've been able to just log from MC

**Comment by ndeadly** [comment]
> Once connected/authenticated in Bluetooth, there should actually really be no difference in handling the controller in the code (except for some differences in the HID reports). So if XBE2 does not connect, it's probably somewhere in the Bluetooth handshake phase

@kakra I've already got the controller IDs in place. I expect everything to just work (bar correct control mappings) as soon as I've got this connection issue figured out. Yeah I'm fairly certain it's the handshake failing somehow. Not sure why it should be any different to the regular XB1 controller though.

**Comment by kakra** [comment]
> Have you got any further info on these privacy extensions? I might try to take a look into it.

Well, there's this section the the Bluez main.conf which needs to be set to `device`:
```
# Default privacy setting.
# Enables use of private address.
# Possible values: "off", "device", "network"
# "network" option not supported currently
# Defaults to "off"
# Privacy = off
Privacy = device
```

Then maybe look at commits to the Linux source code in the directory `net/bluetooth/` which have "privacy" in their commit logs.

**Comment by kakra** [comment]
BTW: That Bluez setting really needs a current kernel. Back when I tried in 5.4, it made no difference. In 5.10 it seems to do its job. I think proper support appeared somewhere around 5.8 or 5.9, so you may look at commits in `{drivers,net}/bluetooth/` if you find anything interesting which may raise an eyebrow or two for you. ;-)

I'm pretty much lost when it comes to understanding how the Bluetooth protocol works. Information about that is either sparse or cryptic. :-)

**Comment by kakra** [comment]
There may be some useful insights here: https://github.com/atar-axis/xpadneo/issues/198

**Comment by kakra** [comment]
> I've already got the controller IDs in place.

Yeah but `0x0B00` probably doesn't belong to that list - it's a USB mode ID, not used in Bluetooth mode by the XBE2 controller.

**Comment by ndeadly** [comment]
> Well, there's this section in the Bluez main.conf which needs to be set to `device`:

Thanks, I'll see if I can figure out what this corresponds to in the Bluetooth spec.

> I'm pretty much lost when it comes to understanding how the Bluetooth protocol works. Information about that is either sparse or cryptic. :-)

Yeah I know what you mean. My understanding of classic Bluetooth is patchy at best. Most of it has come from reading android source. I've been trying to read up on LE in order to implement support for MC but it's spread across so many different documents and difficult to make sense of. Best info I've found has come from nordic semiconductor rather than the spec itself.

> Yeah but `0x0B00` probably doesn't belong to that list - it's a USB mode ID, not used in Bluetooth mode by the XBE2 controller.

Yeah I was just covering all bases. A lot of sources online don't make it obvious which interface the ID s correspond to. I'll likely add USB support in the future anyway

**Comment by kakra** [comment]
> I'll likely add USB support in the future anyway

In that case the controllers no longer speak HID but Microsofts proprietary GIP. For the latter one, you're going to need the init sequence with op-code 0x05 again. GIP is similar in structure but needs sequence numbers, init sequences, and op-codes are different. I'm also planning on implementing GIP support in xpadneo, so we eventually could join forces.

**Comment by ndeadly** [comment]
So recently yellows8 added [this branch](https://github.com/Atmosphere-NX/Atmosphere/tree/uart_mitm) to Atmosphere which can log the low level uart communication between the Bluetooth driver and the hardware. I made a build of it if someone wants to install it and make a log of their attempts to pair the controller. This should yield me something very similar to the android btsnoop logs somebody posted above to compare.

[atmosphere-0.18.0-uart_mitm-1eb4c992-dirty.zip](https://github.com/ndeadly/MissionControl/files/5968951/atmosphere-0.18.0-uart_mitm-1eb4c992-dirty.zip)

Install the above build as you would normally install Atmosphere and add the line `enable_uart_mitm = u8!0x1` under the section `[atmosphere]` in `/atmosphere/config/system_settings.ini` then reboot the console to enable the logging. The logs will be created in `atmosphere/uart_logs`

**Comment by kylefmohr** [comment]
I'm assuming I should have MissionControl installed while I collect these logs, is that correct?

**Comment by ndeadly** [comment]
@kylefmohr yeah, without it running the switch won't even attempt to pair with the controller. You might also want to try with [this build](https://github.com/ndeadly/MissionControl/files/5954437/MissionControl-0.4.1-settsi-mitm.zip) I made a couple of days ago. I discovered a vendor command the switch sends on connection was causing some controllers to be disconnected by the console. Might be worth ruling out the possibility of that messing with anything

**Comment by kylefmohr** [comment]
Hmm, oddly I believe that system_settings.ini addition `enable_uart_mitm = u8!0x1` is causing Atmosphere to panic upon boot. It's blaming Title ID 010041544D530000, which I do not have in my /atmosphere/contents folder. Seems to be a well-documented issue with various causes, but none of the fixes I've seen worked for me. Before I started working on this, I wasn't using a system_settings.ini file, I installed your linked version of Atmosphere, and created the system_settings.ini file with just the single line in it. It panicked, so I deleted the Atmosphere and sept folder, and re-added them to my SD from your zip file. Same panic once again. I then downloaded and installed the latest release of Atmosphere from the official Github release, installed the system_settings.ini file and that new missioncontrol build, same panic. Finally, I deleted the system_settings.ini file and it booted just fine. 

Let me know if there's something else I should try.

tl;dr couldn't collect logs yet

**Comment by ndeadly** [comment]
@kylefmohr that title ID is for the ams_mitm module, which hosts mitms for various services. The cause of the crash could be any number of things so I wouldn't waste too much time trying random fixes. You may have a malformed system_settings.ini if you created it yourself. Try the one I'm using, I created it from the template.

[system_settings.zip](https://github.com/ndeadly/MissionControl/files/5969394/system_settings.zip)

**Comment by kylefmohr** [comment]
Thanks for the insight - I will try yours, but also, I was just typing a response that I just tried a different version of the system_settings.ini file, I copied it from the config_templates folder, and added the uart line to the end of it. I could have sworn I tried that earlier. I'll let you know if I am successful with getting these logs

**Comment by kylefmohr** [comment]
Your settings file did the trick, here are the logs: 
[uart_logs.zip](https://github.com/ndeadly/MissionControl/files/5969431/uart_logs.zip)

**Comment by ndeadly** [comment]
Cool, thanks. Did you do this with other wireless controllers connected? There seems to be a bunch of noise from another device in the log

**Comment by kylefmohr** [comment]
Oh yeah, that was from my Switch Pro controller. One sec, I'll redo it!

**Comment by kylefmohr** [comment]
[uart_logs.zip](https://github.com/ndeadly/MissionControl/files/5969505/uart_logs.zip)

Hopefully this is better, I turned off bluetooth on all nearby devices

**Comment by ndeadly** [comment]
Sorry for the lack of updates guys. @kylefmohr I wasn't really able to spot anything in those uart logs other than what looks like the console disconnecting the controller before authentication completes (which we already knew)

Has anyone attempted to pair the elite v2 on 12.0.1? Nintendo rewrote a bunch of the bluetooth code, so it's _possible_ that their changes allow the controller to complete pairing.

**Comment by DatenThielt** [comment]
> Sorry for the lack of updates guys. @kylefmohr I wasn't really able to spot anything in those uart logs other than what looks like the console disconnecting the controller before authentication completes (which we already knew)
> 
> Has anyone attempted to pair the elite v2 on 12.0.1? Nintendo rewrote a bunch of the bluetooth code, so it's _possible_ that their changes allow the controller to complete pairing.

Fun fact,  If I load MissionControl-0.4.1-settsi-mitm.zip with firmware v12.0.1  the switch refuses to boot! :D

**Comment by ndeadly** [comment]
@DatenThielt this is to be expected. Any builds prior to the 0.5.0 pre-release are compiled with an outdated libstratosphere and will not work with atmosphere 0.19.0+. The changes in that build have since been merged anyway, so there is no reason to use it over the latest release.

**Comment by DatenThielt** [comment]
That would explain that! did not see the pre-release,  I will try with 0.5.0 now!

**Comment by DatenThielt** [comment]
So switch loaded fine,  Went to Controllers and Sensors, Change Controller/Grip Order.  Then entered paired mode on the Elite series 2 controller, nothing happened on the switch at all and the controller is still in pairing mode.  Is there anything you need from me

**Comment by ndeadly** [comment]
You could try it a few more times just to be sure, and maybe check if your controller is on the latest firmware, but that's about all I can think of. I'm not overly hopeful that anything will have changed with regards to this controller working though. Just thought it could be worth a shot since the 12.0.0 firmware update made some reasonably significant changes to the code.

**Comment by DatenThielt** [comment]
Yup thats the plan, It was a new controller today so of course it has some firmware updates,  Doing those as we speak,  I will try it again a few more times after and see if the switch does anything different and report back!

**Comment by DatenThielt** [comment]
No Go unfortunate,  I thought maybe because the controller had no name either it might not work so I gave it a name from the xbox app but still not anything happening on the switch side unfortunately, I'm quite technical and willing to put in any effort to help get this working if needed! (I cant buy you a controller I'm afraid!)

**Comment by ndeadly** [comment]
That's a shame. As far as I can tell from the uart logs posted earlier, the controller _is_ actually being authenticated with the console, but something is happening afterwards that causes it to be disconnected before pairing is considered to be complete. I've been told this controller sends a massive hid report descriptor, so it could be something like this triggering it. I think what really needs to happen is debugging the bluetooth process on-console using something like twili as you try to pair the controller, to try to determine where it actually fails. If this sounds like something you'd be capable of, then I'd be quite happy for you to try it. Unfortunately I can't offer you too much in the way of guidance though.

**Comment by DatenThielt** [comment]
So I did attempt Twili, got it installed, cert installed, drivers installed.  however it seems its not compatible with the latest version of Atmosphere as it crashes atmosphere on startup,  :/

**Comment by ndeadly** [comment]
Yeah I'm not sure if it's been updated or not. Haven't used it in a while. It may be better to wait for the debugging tools that sciresm has been working on. These will be based on Nintendo's own tools so should be a lot easier to work with and kept up to date with atmosphere

**Comment by DatenThielt** [comment]
So ignoring Twili,  Is there any way to debug/breakpoint the code whilst its running on the switch?  Or to get it to pop up any kind of debug messaging other than verbose? Happy to try anything to help :D

**Comment by kakra** [comment]
> I've been told this controller sends a massive hid report descriptor, so it could be something like this triggering it.

The descriptor becomes much smaller after updating the firmware of the controller using the Xbox Accessories app. Also, I've seen it simply disconnect if some Bluetooth packets are sent in some unexpected order but at least in Linux, this disconnect was host-initiated: It would first authenticate successfully, then become disconnected with "authentication failed" which is somewhat strange, it's like the controller is in some quantum fluctuation and when you look at it: boom!

**Comment by ndeadly** [comment]
@kakra this sounds a lot like what we're seeing here actually. How was this resolved under Linux?

**Comment by ndeadly** [comment]
> So ignoring Twili, Is there any way to debug/breakpoint the code whilst its running on the switch? Or to get it to pop up any kind of debug messaging other than verbose? Happy to try anything to help :D

I don't know of any other tools that can be used to debug code running on the switch. As I said earlier, SciresM has been working on a reimplementation of some of Nintendo's own debugging tools. This should end up being a superior alternative to Twili, but currently there is no ETA as to when that might be released.

**Comment by kakra** [comment]
> @kakra this sounds a lot like what we're seeing here actually. How was this resolved under Linux? Or do you mean to say it's intermittent and still occurs?

Still intermittently occurs. Sometimes it takes 10 tries in a row to successfully connect it. Especially the XBE2 controller seems to be sensitive to other Bluetooth devices around, especially while a discovery/scan is running.

**Comment by ndeadly** [comment]
> So ignoring Twili, Is there any way to debug/breakpoint the code whilst its running on the switch? Or to get it to pop up any kind of debug messaging other than verbose? Happy to try anything to help :D

As an update on this - twili has now been updated to work with 12.x.x, but I have also just had luck debugging with gdb using [sys-gdbstub](https://github.com/mossvr/sys-gdbstub) to dig into the running bluetooth process in order to understand and make patches to get Wii controllers pairing again on 12.x.x. It seems as though a similar thing was happening there -  controller pairs successfully according to hci logs but is disconnected before it gets added to the pairing database because something else fails directly afterwards. I'd love to be able to look into this too if I had access to the controller, while it's all fresh in my mind. If you'd like any help setting it up and investigating for yourself I'd be happy to help.

**Comment by XxPowerEyexX** [comment]
i would love to have help, ive been wanting to pair my elite 2 to my switch for a while now

**Comment by kakra** [comment]
> It seems as though a similar thing was happening there - controller pairs successfully according to hci logs but is disconnected before it gets added to the pairing database because something else fails directly afterwards.

This sounds a bit like my issue with some models documented here: https://github.com/bluez/bluez/issues/127. If you can confirm this, it indicates a problem in the controller firmware rather than the Bluetooth stack implementation because we would now have two BT stacks with this problem. In my case, pairing usually works if you try hard enough over a longer period of time: Once in a while, the ordering of the HCI auth events would be different and the controller pairs just fine.

**Comment by ndeadly** [comment]
Looking at that hci log again, I think the authentication part is succeeding every time. It looks like the console is sending the disconnect command after opening the hid connection and receiving a bunch of packet fragments relating to requesting the device service attributes. I'm wondering whether this is exceeding some internal buffer size and  the connection is being terminated because of it. I don't see these fragmented packets with other controllers so it looks like the response is a lot larger.

**Comment by DatenThielt** [comment]
> > So ignoring Twili, Is there any way to debug/breakpoint the code whilst its running on the switch? Or to get it to pop up any kind of debug messaging other than verbose? Happy to try anything to help :D
> 
> As an update on this - twili has now been updated to work with 12.x.x, but I have also just had luck debugging with gdb using [sys-gdbstub](https://github.com/mossvr/sys-gdbstub) to dig into the running bluetooth process in order to understand and make patches to get Wii controllers pairing again on 12.x.x. It seems as though a similar thing was happening there - controller pairs successfully according to hci logs but is disconnected before it gets added to the pairing database because something else fails directly afterwards. I'd love to be able to look into this too if I had access to the controller, while it's all fresh in my mind. If you'd like any help setting it up and investigating for yourself I'd be happy to help.

I will see if I can setup the environment a bit later with sys-gdbstub ,  what spesifcially do you need from this, just an output of it trying to pair?

**Comment by DatenThielt** [comment]
Ok so I have GDB setup and connecting,  When I try and get it to attach to process 102 Bluetooth it just gives me an error (Probably because I am not giving it some arguments it needs?)  

![image](https://user-images.githubusercontent.com/4466326/126186430-75a5a6ad-9542-4780-833f-ecaa4e5d3ed2.png)

**Comment by ndeadly** [comment]
I'm not sure whether you can use regular gdb or not, I use the twili fork https://github.com/misson20000/twili-gdb. I also compiled that with python support (not enabled by default) that enables me to run scripts for doing useful things like creating a hex dump of memory regions

**Comment by DatenThielt** [comment]
> I'm not sure whether you can use regular gdb or not, I use the twili fork https://github.com/misson20000/twili-gdb. I also compiled that with python support (not enabled by default) that enables me to run scripts for doing useful things like creating a hex dump of memory regions

I think the first thing I will need to do is just setup one of my Pi's to run a clean Linux distro as getting it to compile any of this on windows is turning out to be a pain in the back side,  I will get twili running and connecting to the Switch,  Once that is done I will need to use one of the sugested tools to figure out the best place to breakpoint the code?

**Comment by ndeadly** [comment]
I also wrote this python script you might find handy. It automates all the repetitive stuff like attaching to the bluetooth process, computing the base address, and adding breakpoints.

[bt_debug.zip](https://github.com/ndeadly/MissionControl/files/6842305/bt_debug.zip)

And if you add this to your `.gdbinit` file you can have it automatically connect and run the python script every time you run gdb. All you've got to do is add your breakpoint addresses to the script. This saved me a lot of frustration

```
# sys-gdbstub init
target extended-remote 192.168.178.46:10000

source bt_debug.py
```

**Comment by ndeadly** [comment]
> I think the first thing I will need to do is just setup one of my Pi's to run a clean Linux distro as getting it to compile any of this on windows is turning out to be a pain in the back side, I will get twili running and connecting to the Switch, Once that is done I will need to use one of the sugested tools to figure out the best place to breakpoint the code?

I've done all of this from a linux VM and it works quite nicely. I use twili-gdb with sys-gdbstub, you don't have to run twili itself. While I think twili might be slightly superior to sys-gdbstub in terms of features, I've always found it a lot more fiddly to set up.

Then the next step will be to load the bluetooth module in one of those tools yeah. Recent firmwares don't contain any symbols so you'll need to do a bit of initial work to get some things labelled to help find your way around. I can point you in the direction of the callback functions that are likely producing the error status that mission control sees, but ultimately you'd have to work backwards from there and maybe do some RE of your own.

**Comment by DatenThielt** [comment]
Actually I completely forgot about VM's,  That is a better idea for sure,  I will get it set up shortly with Twili-gdb and sys-gdbstub and grab your python script whilst I am at it.  Any help with initial pointing will help for sure. Hopefully not too much RE needed :D I have time but not that much time

**Comment by ndeadly** [comment]
> Any help with initial pointing will help for sure

Sure, just let me know which firmware version you're working with so I can find the correct offsets of the functions in question. It might be better to contact me via discord if you need pointers on how to get up and running, as that's starting to stray a bit from the purpose of this thread.

ndeadly#5317

**Comment by ndeadly** [comment]
So with the help of @DatenThielt and almost a month of spending most of my spare time reverse engineering and poking around deep within the callback hell of the bluetooth stack with gdb, I think we've done what I thought impossible and finally cracked this one.

The root of the issue is the size of the sdp record (primarily due the huge hid report descriptor) sent by this controller, and the fact that stack uses a bunch of fixed-size arrays and buffers defined for the platform and baked in at compile time that are not large enough.

After requesting the sdp record from the controller, a 1000 byte temporary buffer is requested from the memory pool to store the record. The sdp record is around 1300 bytes in length however. and the operation fails upon checking the size (the record from the Xbox One controller is around 590 bytes in comparison). I was able to patch the function to increase the requested buffer size and get the sdp record stored successfully. Victory! Or so I thought...

Next, the code tries to parse the individual attributes from the sdp record into a tree structure and store them in a global database.  When it reaches the hid report descriptor attribute, it fails once again due to not having enough space in the database to store the attribute. Unfortunately, this time the size can't by increased via patching because the memory pool won't allocate us a larger buffer for the database structure.

After many unsuccessful attempts at trying to bypass various checks or ignore the hid report descriptor entirely, I came up with the idea of hooking the attribute parsing function and injecting a minimal dummy descriptor attribute in place of the original. This appears to have done the trick and the controller can now be paired successfully and works flawlessly according to @DatenThielt . As far as I am aware, the hid report descriptor is not actually used for anything outside of being parsed and stored by the bluetooth stack anyway.

Here is a build of v0.5.0 modified to support the larger size of the elite v2 input reports and includes the ips patches to enable successful pairing. Please try it out and let me know if it's working for you or not (and which firmware you're running) as I am obviously not able to test it thoroughly. 

[MissionControl-0.5.0-xbox-elite-2-support-4ab6672.zip](https://github.com/ndeadly/MissionControl/files/7003309/MissionControl-0.5.0-xbox-elite-2-support-4ab6672.zip)

**Comment by tomlube** [comment]
> So with the help of @DatenThielt and almost a month of spending most of my spare time reverse engineering and poking around deep within the callback hell of the bluetooth stack with gdb, I think we've done what I thought impossible and finally cracked this one.
> 
> The root of the issue is the size of the sdp record (primarily due the huge hid report descriptor) sent by this controller, and the fact that stack uses a bunch of fixed-size arrays and buffers defined for the platform and baked in at compile time that are not large enough.
> 
> After requesting the sdp record from the controller, a 1000 byte temporary buffer is requested from the memory pool to store the record. The sdp record is around 1300 bytes in length however. and the operation fails upon checking the size (the record from the Xbox One controller is around 590 bytes in comparison). I was able to patch the function to increase the requested buffer size and get the sdp record stored successfully. Victory! Or so I thought...
> 
> Next, the code tries to parse the individual attributes from the sdp record into a tree structure and store them in a global database. When it reaches the hid report descriptor attribute, it fails once again due to not having enough space in the database to store the attribute. Unfortunately, this time the size can't by increased via patching because the memory pool won't allocate us a larger buffer for the database structure.
> 
> After many unsuccessful attempts at trying to bypass various checks or ignore the hid report descriptor entirely, I came up with the idea of hooking the attribute parsing function and injecting a minimal dummy descriptor attribute in place of the original. This appears to have done the trick and the controller can now be paired successfully and works flawlessly according to @DatenThielt . As far as I am aware, the hid report descriptor is not actually used for anything outside of being parsed and stored by the bluetooth stack anyway.
> 
> Here is a build of v0.5.0 modified to support the larger size of the elite v2 input reports and includes the ips patches to enable successful pairing. Please try it out and let me know if it's working for you or not (and which firmware you're running) as I am obviously not able to test it thoroughly.
> 
> [MissionControl-0.5.0-xbox-elite-2-support-4ab6672.zip](https://github.com/ndeadly/MissionControl/files/7003309/MissionControl-0.5.0-xbox-elite-2-support-4ab6672.zip)

You beautiful man. I can't believe it! Thank you so much.

**Comment by kylefmohr** [comment]
Holy crap! Can confirm it works. Thank you so much for your hard work on this 😄

**Comment by kylefmohr** [comment]
@ndeadly do you have a Bitcoin Cash wallet? If not I'll convert to regular BTC.

**Comment by DatenThielt** [comment]
I think just to solidify how hard this has been

First, lets not sell it short it was just shy of a month, huge achievement and 95% @ndeadly  and 5% me, In this process I have learnt how to read assembly,  debug applications using a decompiler as well as learnt how Bluetooth works, HID descriptors and many other new and horrifying things,  Being in software dev for 10 years helped me understand about 30% of it!  This stuff is crazy hard and once we got into recursive functions (Adding the HID Descriptors) I was 100% useless.   But from my own testing and playing a fair bit of Zelda, I can confirm this is 100% working

I don't know how to state ho hard @ndeadly  worked on this but for a while it was 3-4AM (He was Euro, I am Canada) night of crazy code and head scratching with this bastardized Microsoft device,  So really guys, if you have ever needed a reason to donate or support a dev, He certainly deserves it :)  ( know I will be!)

Now! on to getting HTC Vive controllers working! lmao!

**Comment by ndeadly** [comment]
> @ndeadly do you have a Bitcoin Cash wallet? If not I'll convert to regular BTC. 

No I don't, what's the difference? I can make one if there's any benefit (though it will have to be in the morning as it's kinda late here and I have to be up early for a vaccination appointment)

**Comment by ndeadly** [comment]
> I think just to solidify how hard this has been
> 
> First, lets not sell it short it was just shy of a month, huge achievement and 95% @ndeadly  and 5% me, In this process I have learnt how to read assembly,  debug applications using a decompiler as well as learnt how Bluetooth works, HID descriptors and many other new and horrifying things (Thanks to ndeadly teaching me!),  Being in software dev for 10 years helped me understand about 30% of it!  This stuff is crazy hard and once we got into recursive functions (Adding the HID Descriptors) I was 100% useless.   But from my own testing and playing a fair bit of Zelda, I can confirm this is 100% working
> 
> I don't know how to state how hard @ndeadly  worked on this but for a while it was 3-4AM (He was Euro, I am Canada) night of crazy code and head scratching with this bastardized Microsoft device,  So really guys, if you have ever needed a reason to donate or support a dev, He certainly deserves it :)  ( know I will be!)
> 
> Now! on to getting HTC Vive controllers working! lmao!

Thanks for the kind words, and glad you learned something along the way!

**Comment by kylefmohr** [comment]
> > @ndeadly do you have a Bitcoin Cash wallet? If not I'll convert to regular BTC.
> 
> No I don't, what's the difference? I can make one if there's any benefit (though it will have to be in the morning as it's kinda late here and I have to be up early for a vaccination appointment)

@ndeadly nah don't worry about it, I just use it because it has lower transaction fees, but BTC tx fees are pretty reasonable right now so I'll convert and send some your way shortly here. 

Wish I could do more, as you and @DatenThielt no doubt deserve it! You both have my gratitude.

**Comment by ndeadly** [comment]
Support has been merged into the latest release https://github.com/ndeadly/MissionControl/releases/tag/v0.5.1

I think we can finally lay this issue to rest. Thanks to all of you that helped out over the past year or so.

**Comment by ndeadly** [comment]
Need XBE2 [patches](https://github.com/ndeadly/MissionControl/issues/519#issuecomment-1295659074) for 15.0.0 tested. Can anyone confirm whether the controller can be paired on 15.0.0 with them installed? Thanks.

**Comment by tomlube** [comment]
@ndeadly not working for me on newest MissionControl+Horizon+Elite 2.

**Comment by ndeadly** [comment]
@tomlube non LE firmware? (< 5.x.x)?

**Comment by ndeadly** [comment]
The controller firmware after 5.x.x.x converts the controller to use Bluetooth LE which isn't supported by mission control. You need to downgrade for it to work

**Comment by alint77** [comment]
> Need XBE2 [patches](https://github.com/ndeadly/MissionControl/issues/519#issuecomment-1295659074) for 15.0.0 tested. Can anyone confirm whether the controller can be paired on 15.0.0 with them installed? Thanks.

Hi! just tried it with my XBE2 4.x FW on 15.0.0 with patches and it crashes the CFW when you try to pair with the controller :
<img width="1232" alt="Screen Shot 2022-10-30 at 2 02 05 PM" src="https://user-images.githubusercontent.com/48515733/198874054-58f68a44-d4b6-4f48-b9e3-0b2ec449d533.png">

**Comment by ndeadly** [comment]
@alint77 thanks for the report. I'll have another look over my patches in conjunction with this and see if I can find where I've messed up.

**Comment by ndeadly** [comment]
@alint77 try this

[mc_1500_exefs_patches_2.zip](https://github.com/ndeadly/MissionControl/files/9895787/mc_1500_exefs_patches_2.zip)

**Comment by alint77** [comment]
> @alint77 try this
> 
> [mc_1500_exefs_patches_2.zip](https://github.com/ndeadly/MissionControl/files/9895787/mc_1500_exefs_patches_2.zip)

This time it doesn't crash but wont pair either.
DS4 works as intended

**Comment by ndeadly** [comment]
> > @alint77 try this
> > [mc_1500_exefs_patches_2.zip](https://github.com/ndeadly/MissionControl/files/9895787/mc_1500_exefs_patches_2.zip)
> 
> This time it doesn't crash but wont pair either. DS4 works as intended

That's progress. I had a bug where a couple of branches weren't being corrected. I've looked over my patch code and offsets carefully and can't see anything obviously wrong with the code. Try this one. No actual patch logic was changed, but I swapped the order of the Wii code and the Xbox code so that the Xbox code now pretty much sits in one contiguous piece of memory. Maybe there's a bug in my script for spltting the code across free memory chunks that's messing up the logic.

And please double check the exact controller firmware you have so that neither of us is wasting our time here.

[mc_1500_exefs_patches_3.zip](https://github.com/ndeadly/MissionControl/files/9897044/mc_1500_exefs_patches_3.zip)

**Comment by alint77** [comment]
@ndeadly unfortunately i'm traveling today and can't test it. I'll be back home tomorrow. 

I'm 100% sure about the firmware tho.

**Comment by alint77** [comment]
> > > @alint77 try this
> > > [mc_1500_exefs_patches_2.zip](https://github.com/ndeadly/MissionControl/files/9895787/mc_1500_exefs_patches_2.zip)
> > 
> > 
> > This time it doesn't crash but wont pair either. DS4 works as intended
> 
> That's progress. I had a bug where a couple of branches weren't being corrected. I've looked over my patch code and offsets carefully and can't see anything obviously wrong with the code. Try this one. No actual patch logic was changed, but I swapped the order of the Wii code and the Xbox code so that the Xbox code now pretty much sits in one contiguous piece of memory. Maybe there's a bug in my script for spltting the code across free memory chunks that's messing up the logic.
> 
> And please double check the exact controller firmware you have so that neither of us is wasting our time here.
> 
> [mc_1500_exefs_patches_3.zip](https://github.com/ndeadly/MissionControl/files/9897044/mc_1500_exefs_patches_3.zip)

ok so i just tested the new patch and still no luck.

ds4 pairs but XBE2 doesn't.

triple checked the controller firmware , it is in fact running v4.8.

controller pairs and works with my phone and my macbook.

is there any chance I'm missing something?

**Comment by Banz99** [comment]
> > > > @alint77 try this
> > > > [mc_1500_exefs_patches_2.zip](https://github.com/ndeadly/MissionControl/files/9895787/mc_1500_exefs_patches_2.zip)
> > > 
> > > 
> > > This time it doesn't crash but wont pair either. DS4 works as intended
> > 
> > 
> > That's progress. I had a bug where a couple of branches weren't being corrected. I've looked over my patch code and offsets carefully and can't see anything obviously wrong with the code. Try this one. No actual patch logic was changed, but I swapped the order of the Wii code and the Xbox code so that the Xbox code now pretty much sits in one contiguous piece of memory. Maybe there's a bug in my script for spltting the code across free memory chunks that's messing up the logic.
> > And please double check the exact controller firmware you have so that neither of us is wasting our time here.
> > [mc_1500_exefs_patches_3.zip](https://github.com/ndeadly/MissionControl/files/9897044/mc_1500_exefs_patches_3.zip)
> 
> ok so i just tested the new patch and still no luck.
> 
> ds4 pairs but XBE2 doesn't.
> 
> triple checked the controller firmware , it is in fact running v4.8.
> 
> controller pairs and works with my phone and my macbook.
> 
> is there any chance I'm missing something?

Pairing of Xbox Controllers in general doesn't seem to that immediate on my end, can you try keeping it in pairing mode while on the Controllers->Change Grip/Order for a bit longer?

**Comment by ndeadly** [comment]
Waiting for clarification, but this user claims it's working for them https://github.com/ndeadly/MissionControl/issues/519#issuecomment-1297969144

> is there any chance I'm missing something?

I'm not really sure, I've also checked my patches several times using various methods and I can't see anything wrong or different from previous firmwares. Are you giving the controller plenty of time and trying again it it stops flashing? Xbox controllers can take a bit longer than some to get connected.

Unfortunately all of this stuff takes place deep within the bluetooth stack, so the only true way to debug it is to, well, attach a debugger while the controller is pairing. I don't own one so that's kind of out of the question. You could try my [debug build](https://github.com/ndeadly/MissionControl/files/9836474/MissionControl-0.8.0-debug-logging-7eafa57.zip) of 0.8.0 and see if there's potentially some other failure independent of the actual authenitcation I'm patching. Post the `mc-mitm.log` file this creates on your sd card after attempting to pair the controller.

**Comment by alint77** [comment]
> > > > > @alint77 try this
> > > > > [mc_1500_exefs_patches_2.zip](https://github.com/ndeadly/MissionControl/files/9895787/mc_1500_exefs_patches_2.zip)
> > > > 
> > > > 
> > > > This time it doesn't crash but wont pair either. DS4 works as intended
> > > 
> > > 
> > > That's progress. I had a bug where a couple of branches weren't being corrected. I've looked over my patch code and offsets carefully and can't see anything obviously wrong with the code. Try this one. No actual patch logic was changed, but I swapped the order of the Wii code and the Xbox code so that the Xbox code now pretty much sits in one contiguous piece of memory. Maybe there's a bug in my script for spltting the code across free memory chunks that's messing up the logic.
> > > And please double check the exact controller firmware you have so that neither of us is wasting our time here.
> > > [mc_1500_exefs_patches_3.zip](https://github.com/ndeadly/MissionControl/files/9897044/mc_1500_exefs_patches_3.zip)
> > 
> > 
> > ok so i just tested the new patch and still no luck.
> > ds4 pairs but XBE2 doesn't.
> > triple checked the controller firmware , it is in fact running v4.8.
> > controller pairs and works with my phone and my macbook.
> > is there any chance I'm missing something?
> 
> Pairing of Xbox Controllers in general doesn't seem to that immediate on my end, can you try keeping it in pairing mode while on the Controllers->Change Grip/Order window for a bit longer?

i waited untill controller went into idle mode.

seeing the other thread i just reinstalled Mission Control and retried patch#2 and still no luck... it's extremely odd since my switch is running the exact same version of everything as that person's...

im completely at a loss here

**Comment by ndeadly** [comment]
@alint77 I think the other guy may have been mistaken, he said he cleared his pairing database and now it's not working. I think he might have already had the controller paired and it connected to the console rather than entering pairing mode. 

I'm not really sure where to go from here, as I can't really tell what's going on without debugging on-console. Try that debug build I suggested in the previous post just to rule out some other issue that I _do_ have some kind of control over.

**Comment by corybsa** [comment]
@ndeadly @alint77 I was mistaken. I tried the various patch files and it is not pairing. What debug build are you talking about? I'm also not afraid of poking around the code if there's any way I can help debug this. I've never done any development on the switch but I learn quick.

**Comment by ndeadly** [comment]
@corybsa linked here https://github.com/ndeadly/MissionControl/issues/3#issuecomment-1298708397

It's not really an issue of development, but of debugging the running process in gdb and crossreferencing with the decompilation in IDA or Ghidra. Happy for you try, but it's a lot to have to try and explain to somebody with no familiarity.

**Comment by corybsa** [comment]
@ndeadly That does sound like a lot to explain to a noob. Here's my [mc-mitm.log](https://github.com/ndeadly/MissionControl/files/9913123/mc-mitm.log) file.

**Comment by ndeadly** [comment]
> @ndeadly That does sound like a lot to explain to a noob. Here's my [mc-mitm.log](https://github.com/ndeadly/MissionControl/files/9913123/mc-mitm.log) file.

Yeah, it's just trying to connect over and over with a generic pairing failure code.

Soooo... there _may_ have been an off by one error in my branches. Try this new one
[mc_1500_exefs_patches_4.zip](https://github.com/ndeadly/MissionControl/files/9913185/mc_1500_exefs_patches_4.zip)

**Comment by alint77** [comment]
> > @ndeadly That does sound like a lot to explain to a noob. Here's my [mc-mitm.log](https://github.com/ndeadly/MissionControl/files/9913123/mc-mitm.log) file.
> 
> Yeah, it's just trying to connect over and over with a generic pairing failure code.
> 
> Soooo... there _may_ have been an off by one error in my branches. Try this new one [mc_1500_exefs_patches_4.zip](https://github.com/ndeadly/MissionControl/files/9913185/mc_1500_exefs_patches_4.zip)

Hallelujah! 

it works ! took around 30 seconds to connect... 

i salute you sir 🥇

**Comment by ndeadly** [comment]
Nice! Can one of you confirm that this also works? I just swapped the order of the patches back again
[mc_1500_exefs_patches_5.zip](https://github.com/ndeadly/MissionControl/files/9913334/mc_1500_exefs_patches_5.zip)

And for future me, and anyone else wondering what happened - when installing a hook/detour you need to overwrite an instruction to branch to your injected code. Upon returning to the original function you need to execute this overwritten instruction so as not to change the original code behaviour. In restructuring my patches I had to tweak the way some branch offsets were computed and referred back to my comments that said "jump forward to return statement". While this is technically correct regarding the operation of my code, it ignores the fact that we have this extra instruction to execute prior to returning. Derp.

**Comment by corybsa** [comment]
Just to make sure, I wiped my bluetooth pairing database before installing patch 5, and reverted back to the non-debug build. The controller still successfully pairs.

**Comment by alint77** [comment]
> Nice! Can one of you confirm that this also works? I just swapped the order of the patches back again [mc_1500_exefs_patches_5.zip](https://github.com/ndeadly/MissionControl/files/9913334/mc_1500_exefs_patches_5.zip)
> 
> And for future me, and anyone else wondering what happened - when installing a hook/detour you need to overwrite an instruction to branch to your injected code. Upon returning to the original function you need to execute this overwritten instruction so as not to change the original code behaviour. In restructuring my patches I had to tweak the way some branch offsets were computed and referred back to my comments that said "jump forward to return statement". While this is technically correct regarding the operation of my code, it ignores the fact that we have this extra instruction to execute prior to returning. Derp.

i did the same as @corybsa and it does indeed connect after 20 or so seconds.

**Comment by ndeadly** [comment]
Perfect! I think we can conclude that my patch generation scripts are working correctly now. Should be just a matter of finding the new offsets going forward. Big thanks to both of you 😘

> it does indeed connect after 20 or so seconds.

Strange, I wonder why yours takes so much longer

**Comment by corybsa** [comment]
No problem, glad to help!

I think the pairing time has something to do with the controller. Mine does take longer to pair sometimes and has actually taken a few pairing attempts in the past. But as long as I can use my fancy Xbox controller on my Switch I won't complain!

**Comment by alint77** [comment]
> Perfect! I think we can conclude that my patch generation scripts are working correctly now. Should be just a matter of finding the new offsets going forward. Big thanks to both of you 😘
> 
> > it does indeed connect after 20 or so seconds.
> 
> Strange, I wonder why yours takes so much longer

Thank you for your dedication and this awesome piece of software.
i wish i could somehow help with the code too but this is way out of my depth unfortunately...

as to why it takes longer , i have no idea but the least i can do is to install the debug version and send you the log file if you think it could help.

**Comment by ndeadly** [comment]
No problem. Good testing is important too.

> i have no idea but the least i can do is to install the debug version and send you the log file if you think it could help.

No point. It will be something deeper in the stack that I can't log, probably related to the parameters used by the controller. Any chance of improving it would involve further debugging of the process with the controller attached and likely more patching. Absolutely not worth the time so that someone can save a few seconds waiting for the controller to pair.

**Comment by alint77** [comment]
> No problem. Good testing is important too.
> 
> > i have no idea but the least i can do is to install the debug version and send you the log file if you think it could help.
> 
> No point. It will be something deeper in the stack that I can't log, probably related to the parameters used by the controller. Any chance of improving it would involve further debugging of the process with the controller attached and likely more patching. Absolutely not worth the time so that someone can save a few seconds waiting for the controller to pair.


yes that's true. also i just installed the debug build and it paired much faster , around 4-5 seconds which made me think maybe there's some level of randomness involved , either way its a waste of time.

**Comment by ndeadly** [comment]
It's probably some function of the scan interval used by the console vs the advertise interval used by the controller + other factors. I imagine the interval used console-side is optimised for the switch controllers so I doubt I'd want to mess with that anyway.

**Comment by MrKira34** [comment]
I encountered the same problem; my Xbox One Wireless Controller Elite Series 2.
I installed the latest version of MissionControl 0.11.1. When I try to connect the controller, it just blinks and nothing happens.
Is there a way to connect it?

**Comment by TheTerminatorOfProgramming** [comment]
> I encountered the same problem; my Xbox One Wireless Controller Elite Series 2. I installed the latest version of MissionControl 0.11.1. When I try to connect the controller, it just blinks and nothing happens. Is there a way to connect it?

I am Having this issue as well. i have tried the alpha build for the BLE support and i have confirmed that my elite series 2 is using  firmware version 4.7.1872.0 and has bluetooth support and no BLE support as it is pre 5.x firmware.
